File: photonflow-cpp/src/blocks/measurement/esa_probe.cpp
/**
 * @file esa_probe.cpp
 * @brief ESAProbe implementation.
 */

#include "photonflow/blocks/measurement/esa_probe.hpp"
#include "photonflow/blocks/block_registry.hpp"

namespace photonflow {

ESAProbe::ESAProbe(const std::string &id, const json &params,
                   const json &nonideal)
    : BaseBlock(id, params, nonideal) {}

std::unordered_map<std::string, Signal>
ESAProbe::process(const std::unordered_map<std::string, Signal> &inputs,
                  SimContext & /*ctx*/) {
  auto it = inputs.find("elec_in");
  if (it == inputs.end()) {
    return {};
  }
  return {};
}

std::optional<std::string> ESAProbe::port_type(const std::string &port) const {
  if (port == "elec_in")
    return "electrical";
  return std::nullopt;
}

json ESAProbe::describe() const {
  return {{"ports", {{"elec_in", "electrical"}}},
          {"spec",
           {{"params",
             {{"window",
               {{"type", "enum"},
                {"default", "hann"},
                {"options", json::array({"hann", "hamming", "blackman", "rect",
                                         "kaiser"})}}},
              {"ref", {{"type", "float"}, {"default", 1.0}, {"unit", ""}}},
              {"include_power", {{"type", "bool"}, {"default", false}}}}},
            {"nonideal", json::object()}}}};
}

REGISTER_BLOCK(ESAProbe, "ESAProbe");

} // namespace photonflow

File: backend/src/photonflow/blocks/measurement/esa_probe.py
"""Electrical spectrum analyzer probe."""

from __future__ import annotations

from typing import Dict

from photonflow.blocks.base import BaseBlock, register_block
from photonflow.core.signal import Signal
from photonflow.core.sim import SimContext


@register_block("ESAProbe")
class ESAProbe(BaseBlock):
    """Electrical Spectrum Analyzer (ESA) probe. Measures the power spectrum of an electrical signal, displaying power vs. frequency."""

    PORTS = {"elec_in": "electrical"}
    SPEC = {
        "params": {
            "window": {
                "type": "enum",
                "default": "hann",
                "options": ["hann", "hamming", "blackman", "rect", "kaiser"],
            },
            "ref": {"type": "float", "default": 1.0, "unit": ""},
            "include_power": {"type": "bool", "default": False},
        },
        "nonideal": {},
    }

    def process(self, inputs: Dict[str, Signal], ctx: SimContext) -> Dict[str, Signal]:
        _ = inputs.get("elec_in")
        return {}
